testsuites/syncgateway/upgrade/testserver/test_upgrade.py

```diff
@@ -0,0 +1,424 @@
+    authenticator = Authenticator(base_url)
+    replicator_authenticator = authenticator.authentication(sg_session, sg_cookie, authentication_type="session")
+
+    repl_config = replicator.configure(cbl_db, sg_blip_url, continuous=True, channels=sg_user_channels, replication_type="push_pull", replicator_authenticator=replicator_authenticator)
+    repl = replicator.create(repl_config)
+    replicator.start(repl)
+    replicator.wait_until_replicator_idle(repl)
+
+    # start updating docs
+    terminator_doc_id = 'terminator'
+
+    with ProcessPoolExecutor() as up:
+        # Start updates in background process
+        updates_future = up.submit(
+            update_docs,
+            db,
+            cbl_db,
+            added_docs,
+            doc_obj,
+            terminator_doc_id
+        )
+
+        # Supported upgrade process
+        # 1. Upgrade SGs first docmeta -> docmeta - CBS 5.0.0 does not support TAP.
+        # 2. Upgrade the CBS cluster.
+        # 3. Enable import/xattrs on SGs
+
+        # Upgrade SG docmeta -> docmeta
+        cluster_util = ClusterKeywords(cluster_config)
+        topology = cluster_util.get_cluster_topology(cluster_config, lb_enable=False)
+        sync_gateways = topology["sync_gateways"]
+
+        upgrade_sync_gateway(
+            sync_gateways,
+            sync_gateway_version,
+            sync_gateway_upgraded_version,
+            sg_conf,
+            cluster_config
+        )
+
+        # Upgrade CBS
+        cluster = Cluster(config=cluster_config)
+        if len(cluster.servers) < 2:
+            raise Exception("Please provide at least 3 servers")
+
+        server_urls = []
+        for server in cluster.servers:
+            server_urls.append(server.url)
+
+        primary_server = cluster.servers[0]
+        secondary_server = cluster.servers[1]
+        servers = cluster.servers[1:]
+
+        upgrade_server_cluster(
+            servers,
+            primary_server,
+            secondary_server,
+            server_version,
+            server_upgraded_version,
+            server_urls,
+            cluster_config,
+            cbs_platform,
+            toy_build=cbs_toy_build
+        )
+
+        # Restart SGs after the server upgrade
+        sg_obj = SyncGateway()
+        for sg in sync_gateways:
+            sg_ip = host_for_url(sg["admin"])
+            log_info("Restarting sync gateway {}".format(sg_ip))
+            sg_obj.restart_sync_gateways(cluster_config=cluster_config, url=sg_ip)
+            time.sleep(5)
+
+        if need_to_redeploy:
+            # Enable xattrs on all SG/SGAccel nodes
+            # cc - Start 1 SG with import enabled, all with XATTRs enabled
+            #    - Do not enable import in SG.
+            if mode == "cc":
+                enable_import = True
+
+            sg_obj = SyncGateway()
+            for sg in sync_gateways:
+                sg_ip = host_for_url(sg["admin"])
+                sg_obj.redeploy_sync_gateway_config(
```

Comment:
can you get the status whether sync_gateway successfully started or fail. If it fails, you can fail the test.